packages/Kokkos/src/DTK_UserFunctionRegistry.hpp

```diff
@@ -0,0 +1,319 @@
+  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+//---------------------------------------------------------------------------//
+/*!
+ * \file DTK_UserFunctionRegistry.hpp
+ * \brief Registry for user functions.
+ */
+//---------------------------------------------------------------------------//
+
+#ifndef DTK_USERFUNCTIONREGISTRY_HPP
+#define DTK_USERFUNCTIONREGISTRY_HPP
+
+#include "DTK_ConfigDefs.hpp"
+#include "DTK_DBC.hpp"
+#include "DTK_UserDataInterface.hpp"
+#include "DTK_View.hpp"
+
+#include <functional>
+#include <memory>
+#include <string>
+#include <unordered_map>
+
+namespace DataTransferKit
+{
+
+//---------------------------------------------------------------------------//
+//! Use this namespace for convenience.
+using namespace UserDataInterface;
+
+//---------------------------------------------------------------------------//
+// Forward declaration of UserApplication.
+template <class Scalar, class ParallelModel>
+class UserApplication;
+
+//---------------------------------------------------------------------------//
+/*!
+ * \class UserFunctionRegistry
+ *
+ * \brief Registry for user functions.
+ *
+ * \tparam Scalar Scalar type of the fields this user function implementaton
+ * will provide.
+ *
+ * The registry is the mechanism by which user functions and the data to be
+ * called with those user functions are registered with DTK.
+ */
+//---------------------------------------------------------------------------//
+template <class Scalar>
+class UserFunctionRegistry
+{
+    //! UserApplication is a friend class of the registry to provide direct
+    //! access to user functions. Only UserApplications with the equivalent
+    //! scalar type will access the data of this class. Partial specialization
+    //! of friends is prohibited so we fully declare the template parameters
+    //! here. Because user functions in the registry are private data and have
+    //! no accessors this indicates that the UserApplication class is the only
+    //! object allowed to call them.
+    template <class UserScalarType, class ParallelModel>
+    friend class UserApplication;
+
+  public:
+    //! User implementation.
+    template <class CallableObject>
+    using UserImpl = std::pair<CallableObject, std::shared_ptr<void>>;
+
+    //! Field function map.
+    template <class CallableObject>
+    using UserImplMap =
+        std::unordered_map<std::string, UserImpl<CallableObject>>;
+
+    //@{
+    //! Register geometry.
+
+    // Node list size function.
+    void
+    registerNodeListSizeFunction( NodeListSizeFunction &&func,
```

Comment:
What about set instead of register as the prefix for these functions? It would reduce typing and mean the same thing.